src/NHibernate.Test/TypesTest/DateTimeTypeFixture.cs

```diff
@@ -37,21 +47,296 @@ public void DeepCopyNotNull()
 			object value1 = DateTime.Now;
 			object value2 = type.DeepCopy(value1, null);
 
-			Assert.AreEqual(value1, value2, "Copies should be the same.");
+			Assert.That(value2, Is.EqualTo(value1), "Copies should be the same.");
 
-
-			value2 = ((DateTime)value2).AddHours(2);
-			Assert.IsFalse(value1 == value2, "value2 was changed, value1 should not have changed also.");
+			value2 = ((DateTime) value2).AddHours(2);
+			Assert.That(value2, Is.Not.EqualTo(value1), "value2 was changed, value1 should not have changed also.");
 		}
 
 		[Test]
 		public void EqualityShouldIgnoreKindAndMillisecond()
 		{
-			var type = (DateTimeType)NHibernateUtil.DateTime;
+			var type = NHibernateUtil.DateTime;
 			var localTime = DateTime.Now;
-			var unspecifiedKid = new DateTime(localTime.Year, localTime.Month, localTime.Day, localTime.Hour, localTime.Minute, localTime.Second, 0, DateTimeKind.Unspecified);
+			var unspecifiedKid = new DateTime(
+				localTime.Year,
+				localTime.Month,
+				localTime.Day,
+				localTime.Hour,
+				localTime.Minute,
+				localTime.Second,
+				0,
+				DateTimeKind.Unspecified);
 			Assert.That(type.IsEqual(localTime, unspecifiedKid), Is.True);
 			Assert.That(type.IsEqual(localTime, unspecifiedKid), Is.True);
 		}
 	}
-}
\ No newline at end of file
+
+	[TestFixture]
+	public class DateTimeSqlTypeFixture : TypeFixtureBase
+	{
+		protected override string TypeName => "DateTime";
+		private const int _dateId = 1;
+
+		protected override void Configure(Configuration configuration)
+		{
+			base.Configure(configuration);
+
+			var driverClass = ReflectHelper.ClassForName(configuration.GetProperty(Cfg.Environment.ConnectionDriver));
+			ClientDriverWithParamsStats.DriverClass = driverClass;
+
+			configuration.SetProperty(
+				Cfg.Environment.ConnectionDriver,
+				typeof(ClientDriverWithParamsStats).AssemblyQualifiedName);
+		}
+
+		protected override void OnSetUp()
+		{
+			base.OnSetUp();
+
+			using (var s = OpenSession())
+			using (var t = s.BeginTransaction())
+			{
+				var d = new DateTimeClass
+				{
+					Id = _dateId,
+					LocalDateTimeValue = DateTime.Now.AddDays(-1),
+					UtcDateTimeValue = DateTime.UtcNow.AddDays(-1),
+					NormalDateTimeValue = DateTime.Now.AddDays(-1)
+				};
+				s.Save(d);
+				t.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			base.OnTearDown();
+
+			using (var s = OpenSession())
+			using (var t = s.BeginTransaction())
+			{
+				s.CreateQuery("delete from DateTimeClass").ExecuteUpdate();
+				t.Commit();
+			}
+		}
+
+		[Test]
+		public void DbHasExpectedType()
+		{
+			var validator = new SchemaValidator(cfg);
+			validator.Validate();
```

Comment:
Delegating to `Validate` for checking we have the expected type in database. Which means being confident `Validate` correctly knows if it should be `DateTime` or `DateTime2`.
It appeared to me overly complicated to test this manually for all dialects, so I delegate to `Validate`. I have furthermore checked in a debug session directly in the db what type it was to ascertain we were now having datetime2 with SQL2008.